File: Nao/robot.py
import qi

class Robot:
    def __init__(self, ip: str, port: int = 9559):
        self.ip = ip
        self.port = port
        self.session = self._connect()

    def _connect(self):
        session = qi.Session()
        try:
            session.connect(f"tcp://{self.ip}:{self.port}")
        except RuntimeError as rte:
            if "The call request could not be handled" in str(rte):
                return None
            raise rte
        return session

    def get_service(self, name: str):
        if self.session is None:
            return None
        return self.session.service(name)

File: src/robot.cpp
//
// Created by crystal on 7/25/25.
//

#include "robot.h"

std::shared_ptr<qi::Session> Robot::_connect() const {
	auto session = std::make_shared<qi::Session>();
	try {
		session->connect("tcp://" + ip_ + ":" + std::to_string(port_));
	} catch (const std::runtime_error& e) {
		std::string msg = e.what();
		if (msg.find("The call request could not be handled") != std::string::npos) {
			return nullptr;
		}
		throw;
	}
	return session;
}

qi::AnyObject Robot::getService(const std::string& name) const {
	if (!session_ || !session_->isConnected()) {
		return {};
	}
	return session_->service(name).value();
}
